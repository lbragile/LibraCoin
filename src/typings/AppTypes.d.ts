import { ACTIONS } from "../enums/AppDispatchActions";

export interface ITransaction {
  to: string;
  from: string;
  amount: number | undefined;
  message: string;
  signature: string;
}

export interface IUser {
  publicKey: string;
  balance: number;
}

export interface IBlock {
  index: number;
  prevHash: string;
  currHash: string;
  transactions: ITransaction[];
  timestamp: number;
  merkleRoot: string;
  valid?: boolean;
}
export interface IState {
  verifiedTrans: ITransaction[];
  selectedTrans: ITransaction[];
  users: IUser[];
  chain: IBlock[];
}

export interface IAction {
  type: ACTIONS;
  payload?: unknown;
}
