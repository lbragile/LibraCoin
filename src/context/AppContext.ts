import React from "react";
import { IAction, IState } from "../typings/AppTypes";

type TAppContext = { state: IState; dispatch: React.Dispatch<IAction> } | undefined;

export const AppContext = React.createContext<TAppContext>(undefined);
