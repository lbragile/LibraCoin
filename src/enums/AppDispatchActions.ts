export enum ACTIONS {
  ADD_VERIFIED_TRANS,
  UPDATE_VERIFIED_TRANS,
  UPDATE_SELECTED_TRANS,
  UPDATE_USERS,
  SET_MAIN_USER,
  ADD_BLOCK,
  UPDATE_BLOCK
}
