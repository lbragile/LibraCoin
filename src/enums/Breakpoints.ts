// https://www.w3schools.com/css/css_rwd_mediaqueries.asp

export enum SIZES {
  xs = "600px",
  sm = "600px",
  md = "768px",
  lg = "992px",
  xl = "1200px"
}

export enum DEVICES {
  xs = "max-width: " + SIZES.xs,
  sm = "min-width: " + SIZES.sm,
  md = "min-width: " + SIZES.md,
  lg = "min-width: " + SIZES.lg,
  xl = "min-width: " + SIZES.xl
}
