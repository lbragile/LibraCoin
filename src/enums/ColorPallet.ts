export enum COLORS {
  DEFAULT_BACKGROUND = "#fff",

  DARK_BACKGROUND = "#1f1f1f",

  VALID_BACKGROUND = "#b4ff9a",
  INVALID_BACKGROUND = "#ff9a9a",
  VALID_SOLUTION = "#008000",
  INVALID_SOLUTION = "#f00",

  NAVBAR_BACKGROUND = "#eee",
  NAV_LINK_UNDERSCORE = "#8ac0ff",
  NAV_TEXT = "#000",

  SELECTED_TRANS_BACKGROUND = "#dfd",
  SELECTED_TRANS_BORDER = "#090",

  BLOCK_TRANS_BACKGROUND = "#c7e4ff",

  ROOT_CELL = "#4f9249",
  MIDDLE_CELLS = "#cf8b31",
  DATA_CELLS = "#7d95ff",
  PREVIEW_LINES = "#777"
}
