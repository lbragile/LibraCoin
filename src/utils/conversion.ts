import { TextEncoder } from "text-encoding";

export function bufferToHex(hashBuffer: ArrayBuffer): string {
  const hashArray = Array.from(new Uint8Array(hashBuffer)); // convert buffer to byte array
  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join(""); // convert bytes to hex string
  return hashHex;
}

export async function CryptoKeyToHex(format: string, key: CryptoKey): Promise<string> {
  const buf = (await window.crypto.subtle.exportKey(format, key)) as ArrayBuffer;
  return bufferToHex(buf);
}

export async function digestMessage(message: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
  const hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8); // hash the message
  return bufferToHex(hashBuffer);
}

export function randomHash(len: number): string {
  return bufferToHex(window.crypto.getRandomValues(new Uint32Array(len)));
}
