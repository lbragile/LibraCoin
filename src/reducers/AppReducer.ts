import { ACTIONS } from "../enums/AppDispatchActions";
import { IAction, IState, IUser, ITransaction, IBlock } from "../typings/AppTypes";

export const AppReducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case ACTIONS.ADD_VERIFIED_TRANS: {
      const newTrans = (action.payload as { trans: ITransaction }).trans;
      const prevTrans = JSON.parse(localStorage.getItem("verTrans") as string) ?? [];
      const verifiedTrans = [...prevTrans, newTrans];
      localStorage.setItem("verTrans", JSON.stringify(verifiedTrans));
      return { ...state, verifiedTrans };
    }

    case ACTIONS.UPDATE_VERIFIED_TRANS: {
      const selectedTransSignatures = state.selectedTrans.map((x) => x.signature);
      const verifiedTrans = state.verifiedTrans.filter((x) => !selectedTransSignatures.includes(x.signature));
      localStorage.setItem("verTrans", JSON.stringify(verifiedTrans));
      return { ...state, verifiedTrans };
    }

    case ACTIONS.UPDATE_SELECTED_TRANS: {
      const selectedTrans = (action.payload as { selectedTrans: ITransaction[] }).selectedTrans;
      localStorage.setItem("selTrans", JSON.stringify(selectedTrans));
      return { ...state, selectedTrans };
    }

    case ACTIONS.UPDATE_USERS: {
      const users = (action.payload as { users: IUser[] }).users;
      localStorage.setItem("users", JSON.stringify(users));
      return { ...state, users };
    }

    case ACTIONS.ADD_BLOCK: {
      const { block } = action.payload as { block: IBlock };
      const chain = [...state.chain, block];
      localStorage.setItem("chain", JSON.stringify(chain));
      return { ...state, chain };
    }

    case ACTIONS.UPDATE_BLOCK: {
      const { block } = action.payload as { block: IBlock };
      const chain = JSON.parse(localStorage.getItem("chain") as string);
      chain[block.index] = block;
      localStorage.setItem("chain", JSON.stringify(chain));
      return { ...state, chain };
    }

    default:
      return state;
  }
};
