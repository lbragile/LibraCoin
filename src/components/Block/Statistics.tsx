import React, { useState, useRef, useContext } from "react";
import { Button, Form, InputGroup } from "react-bootstrap";
import { AppContext } from "../../context/AppContext";
// import { ACTIONS } from "../../enums/AppDispatchActions";
import { IAction, IState } from "../../typings/AppTypes";
import { mine } from "../../utils/mine";

interface IStats {
  chain: boolean;
  isValid: boolean;
  solution: string;
  setIsValid: (arg: boolean) => void;
  setSolution: (arg: string) => void;
  index?: number;
}

export default function Statistics(props: IStats): JSX.Element {
  const { state } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  const nonce = useRef<number>(Math.round(Math.random() * 1e6));
  const [header, setHeader] = useState<number>();
  const [target, setTarget] = useState<string>();

  return (
    <div className={props.chain ? "bordered-background" : "col-11 col-lg-5 mx-3"}>
      <InputGroup className="my-2">
        <InputGroup.Prepend>
          <InputGroup.Text>Nonce</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control type="number" defaultValue={props.solution ? nonce.current : ""} disabled={true} />
      </InputGroup>

      <InputGroup className="my-2">
        <InputGroup.Prepend>
          <InputGroup.Text>Header</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control type="number" defaultValue={header} disabled={true} />
      </InputGroup>

      <InputGroup className="my-2">
        <InputGroup.Prepend>
          <InputGroup.Text className="font-weight-bold">Target</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control type="text" defaultValue={target} disabled={true} />
      </InputGroup>

      <InputGroup className="my-2">
        <InputGroup.Prepend>
          <InputGroup.Text>Sol&apos;n</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control
          type="text"
          style={props.isValid ? { color: "green" } : { color: "red" }}
          defaultValue={props.solution}
          disabled={true}
        />
      </InputGroup>

      <Button
        variant="primary"
        className="btn-block d-block mt-3"
        disabled={props.isValid || state.selectedTrans.length === 0}
        onClick={() => mine(nonce.current, setHeader, setTarget, props.setSolution, props.setIsValid)}
      >
        <h4 className="m-0">Mine</h4>
      </Button>
    </div>
  );
}
