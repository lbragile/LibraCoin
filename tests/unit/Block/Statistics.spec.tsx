import React, { useReducer } from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";

import Statistics from "../../../src/components/Block/Statistics";
import { AppContext } from "../../../src/context/AppContext";
import { IAction, IState } from "../../../src/typings/AppTypes";
import { ACTIONS } from "../../../src/enums/AppDispatchActions";
import * as ConversionUtil from "../../../src/utils/conversion";
import { AppReducer } from "../../../src/reducers/AppReducer";

const { initialState } = global;

interface IStatisticsWrapper {
  chain: boolean;
  stateMock?: IState;
  dispatchMock?: React.Dispatch<IAction>;
}

const StatisticsWrapper = ({ chain, stateMock, dispatchMock }: IStatisticsWrapper) => {
  const [state, dispatch] = useReducer(AppReducer, initialState);

  return (
    <AppContext.Provider value={{ state: stateMock ?? state, dispatch: dispatchMock ?? dispatch }}>
      <Statistics chain={chain} />
    </AppContext.Provider>
  );
};

describe("in preview mode", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<StatisticsWrapper chain={false} />);

    expect(screen.getByRole("form", { name: /Block Statistics/i })).toHaveFormValues({
      nonce: 0,
      header: 0,
      target: "",
      solution: ""
    });

    const nonce = screen.getByRole("spinbutton", { name: /Block Nonce/i });
    expect(nonce).toBeDisabled();
    expect(nonce).not.toBeRequired();

    const header = screen.getByRole("spinbutton", { name: /Block Header/i });
    expect(header).toBeDisabled();
    expect(header).not.toBeRequired();

    const target = screen.getByRole("textbox", { name: /Block Target/i });
    expect(target).toHaveAttribute("readOnly");
    expect(target).not.toBeRequired();

    const solution = screen.getByRole("textbox", { name: /Block Solution/i });
    expect(solution).toHaveAttribute("readOnly");
    expect(solution).not.toBeRequired();
    expect(solution).toHaveClass("invalid-solution");

    const mineBtn = screen.getByRole("button", { name: /Block Mine/i });
    expect(mineBtn).toHaveTextContent("Mine");
    expect(mineBtn).toBeEnabled();

    expect(asFragment()).toMatchSnapshot();
  });

  describe("mining process", () => {
    describe("selected transactions", () => {
      test("that mine button is disabled when there are no selected transactions", () => {
        const ogState: IState = JSON.parse(JSON.stringify(initialState));
        ogState.selectedTrans = [];

        render(<StatisticsWrapper chain={false} stateMock={ogState} />);

        expect(screen.getByRole("button", { name: /Block Mine/i })).toBeDisabled();
      });

      test("that mine button is enabled when there are a few selected transactions", () => {
        render(<StatisticsWrapper chain={false} />);

        expect(screen.getByRole("button", { name: /Block Mine/i })).toBeEnabled();
      });
    });

    describe("button behavior during mining (disabled the button while mining)", () => {
      beforeEach(() => {
        jest.spyOn(ConversionUtil, "randomHash").mockReturnValueOnce("random");
      });

      afterEach(() => {
        jest.restoreAllMocks();
      });

      it("calls dispatch & updates solution/target fields correctly", async () => {
        const badSolution = "032a4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";
        const solution = "000a4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";
        const target = "000z4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";

        const dispatchMock = jest.fn();
        Date.now = jest.fn().mockReturnValueOnce(12345);
        jest
          .spyOn(ConversionUtil, "digestMessage")
          .mockResolvedValueOnce(target)
          .mockResolvedValueOnce(badSolution)
          .mockResolvedValueOnce(solution);

        render(<StatisticsWrapper chain={false} dispatchMock={dispatchMock} />);

        expect(screen.getByRole("button", { name: /Block Mine/i })).toBeEnabled();
        fireEvent.click(screen.getByRole("button", { name: /Block Mine/i })); // mining is async so setting mine button to disabled happens outside React's call stack
        await waitFor(() => expect(screen.getByRole("button", { name: /Block Mine/i })).toBeDisabled());

        expect(screen.getByRole("textbox", { name: /Block Target/i })).toHaveValue(target);
        expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveValue(solution);

        expect(dispatchMock).toHaveBeenCalledTimes(1);
        expect(dispatchMock).toHaveBeenCalledWith({
          type: ACTIONS.UPDATE_PREVIEW,
          payload: {
            preview: {
              ...initialState.preview,
              timestamp: 12345,
              prevHash: initialState.chain[initialState.preview.index - 1].currHash,
              currHash: solution,
              valid: true
            }
          }
        });
      });

      it("enables mining button after mining due to invalid solution", async () => {
        const solution = "000a4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";
        const target = "00034fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";

        jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValueOnce(target).mockResolvedValueOnce(solution);

        render(<StatisticsWrapper chain={false} />);

        expect(screen.getByRole("button", { name: /Block Mine/i })).toBeEnabled();

        // mining is async so setting mine button to disabled happens outside React's call stack
        fireEvent.click(screen.getByRole("button", { name: /Block Mine/i }));

        await waitFor(() => expect(screen.getByRole("button", { name: /Block Mine/i })).toBeDisabled());

        // solution is invalid, so it will re-enable
        await waitFor(() => expect(screen.getByRole("button", { name: /Block Mine/i })).toBeEnabled());
        expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("invalid-solution");
      });

      it("keeps mining button disabled after mining due to valid solution", async () => {
        const solution = "000a4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";
        const target = "000b4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";

        jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValueOnce(target).mockResolvedValueOnce(solution);

        render(<StatisticsWrapper chain={false} />);

        expect(screen.getByRole("button", { name: /Block Mine/i })).toBeEnabled();

        fireEvent.click(screen.getByRole("button", { name: /Block Mine/i }));

        // need to await state changes
        await waitFor(() => expect(screen.getByRole("button", { name: /Block Mine/i })).toBeDisabled());
        expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("valid-solution");
      });
    });
  });
});
