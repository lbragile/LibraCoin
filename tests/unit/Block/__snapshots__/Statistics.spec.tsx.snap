// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in blockchain mode renders correctly block is invalid (index: 2) 1`] = `
<DocumentFragment>
  <form
    aria-label="Block Statistics"
    class="my-3"
  >
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Nonce
        </span>
      </div>
      <input
        aria-label="Block Nonce"
        class="form-control"
        disabled=""
        name="nonce"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Header
        </span>
      </div>
      <input
        aria-label="Block Header"
        class="form-control"
        disabled=""
        name="header"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Target
        </span>
      </div>
      <input
        aria-label="Block Target"
        class="text-truncate form-control"
        name="target"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Sol'n
        </span>
      </div>
      <input
        aria-label="Block Solution"
        class="text-truncate invalid-solution form-control"
        name="solution"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Block Mine"
      class="btn-block mt-2 py-2 btn btn-primary"
      type="button"
    >
      <h4
        class="my-0 d-flex align-items-center justify-content-end"
      >
        Mine
        <span
          class="spinner-border spinner-border-md mx-5 invisible"
          role="status"
        />
      </h4>
    </button>
  </form>
</DocumentFragment>
`;

exports[`in blockchain mode renders correctly block is valid (index: 0) 1`] = `
<DocumentFragment>
  <form
    aria-label="Block Statistics"
    class="my-3"
  >
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Nonce
        </span>
      </div>
      <input
        aria-label="Block Nonce"
        class="form-control"
        disabled=""
        name="nonce"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Header
        </span>
      </div>
      <input
        aria-label="Block Header"
        class="form-control"
        disabled=""
        name="header"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Target
        </span>
      </div>
      <input
        aria-label="Block Target"
        class="text-truncate form-control"
        name="target"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Sol'n
        </span>
      </div>
      <input
        aria-label="Block Solution"
        class="text-truncate valid-solution form-control"
        name="solution"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Block Mine"
      class="btn-block mt-2 py-2 btn btn-primary"
      disabled=""
      type="button"
    >
      <h4
        class="my-0 d-flex align-items-center justify-content-end"
      >
        Mine
        <span
          class="spinner-border spinner-border-md mx-5 invisible"
          role="status"
        />
      </h4>
    </button>
  </form>
</DocumentFragment>
`;

exports[`in preview mode renders correctly 1`] = `
<DocumentFragment>
  <form
    aria-label="Block Statistics"
    class="col-11 col-lg-5 mx-3"
  >
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Nonce
        </span>
      </div>
      <input
        aria-label="Block Nonce"
        class="form-control"
        disabled=""
        name="nonce"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Header
        </span>
      </div>
      <input
        aria-label="Block Header"
        class="form-control"
        disabled=""
        name="header"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Target
        </span>
      </div>
      <input
        aria-label="Block Target"
        class="text-truncate form-control"
        name="target"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Sol'n
        </span>
      </div>
      <input
        aria-label="Block Solution"
        class="text-truncate invalid-solution form-control"
        name="solution"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Block Mine"
      class="btn-block mt-2 py-2 btn btn-primary"
      type="button"
    >
      <h4
        class="my-0 d-flex align-items-center justify-content-end"
      >
        Mine
        <span
          class="spinner-border spinner-border-md mx-5 invisible"
          role="status"
        />
      </h4>
    </button>
  </form>
</DocumentFragment>
`;
